chore: Follow-up Documentation cleanup (#159)
SAP/ai-sdk-js: packages/orchestration/README.md

## Table of Contents

- [Installation](#installation)
- [Prerequisites](#prerequisites)
- [Orchestration Service](#orchestration-service)
- [Relationship between Orchestration and Resource Groups](#relationship-between-orchestration-and-resource-groups)
- [Usage](#usage)
  - [Templating](#templating)
  - [Content Filtering](#content-filtering)
  - [Data Masking](#data-masking)
  - [Retrieving Data from the Response](#retrieving-data-from-the-response)
    - [Finish Reason](#finish-reason)
    - [Token Usage](#token-usage)

>     - [Using Resource Groups](#using-resource-groups)

Author: Finish reason, token usage are meant to be under Retrieving Data from the Response, as they explain about more information from the response. But as other readmes(in other packages like for e.g langchain) don't link minor subsections, I am removing it from the list.

Using Resource groups, should be moved up, have done that.

4. [Usage](#usage)
   - [Templating](#templating)
   - [Content Filtering](#content-filtering)
   - [Retrieving Data from the Response](#retrieving-data-from-the-response)
     - [Finish Reason](#finish-reason)
     - [Token Usage](#token-usage)
     - [Using Resource Groups](#using-resource-groups)

> 5. [Local Testing](#local-testing)

Reviewer: I think we should not use a numbered list in ToC at all. I general list should be fine. We run the risk of getting the wrong number ordering.

Find more details about orchestration workflow [here](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/orchestration-workflow).

## Relationship between Orchestration and Resource Groups

SAP AI Core manages access to orchestration of generative AI models through the global AI scenario `orchestration`.
Creating a deployment for enabling orchestration capabilities requires access to this scenario.

[Resource groups](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/resource-groups?q=resource+group) represent a virtual collection of related resources within the scope of one SAP AI Core tenant.
Each resource group allows for a one-time orchestration deployment.

> Consequently, each orchestration deployment uniquely maps to a resource group within the `orchestration` scenario.

Author: As, the deployment requirements change between orchestration and foundation-models. I would prefer to write about it in the dedicated client document  instead of a separate file, as this would be more confusing.

Find more details about orchestration workflow [here](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/orchestration-workflow).

## Relationship between Orchestration and Resource Groups

Access to orchestration of generative AI models is provided under the global AI scenario `orchestration`, which is managed by SAP AI Core.
A deployment can be created to enable orchestration capabilities only with access to the global AI scenario `orchestration`.
[Resource groups](https://help.sap.com/docs/sap-ai-core/sap-ai-core-service-guide/resource-groups?q=resource+group) represent a virtual collection of related resources within the scope of one SAP AI Core tenant.
Each resource group allows for a one-time orchestration deployment.

> Consequently, each orchestration deployment uniquely maps to a resource group within the `orchestration` scenario.

Reviewer: [pp] Maybe also link the other section where we tell about fetching using resource group

Author: As we haven't explained how to initialise the client yet, I wouldn't link the fetching section here.
I can instead link this in the fetching section.

const responseContent = response.getContent();
```

> #### Passing a custom request configuration

Reviewer 2: ```suggestion
#### Customizing the Request Configuration
```

> To pass custom request configuration, for example additional headers to be passed to the client, use the following snippet:

Reviewer 2: ```suggestion
To customize the request configuration, for example if you want to pass additional headers to the client, use the second parameter in the `chatCompletion()` method:
```

Reviewer 2: [pp] Maybe also add a sentence that this configuration takes precedence over automatically created configuration for your request. 
Also, once there is more information on deployment resolution, make sure to mention that the request config is not used for deployment resolution.

```ts
const response = await orchestrationClient.chatCompletion(
  {
    inputParams: { country: 'France' }
  },
  {
    headers: { 'x-custom-header': 'custom-value' }
  }
);
```

> #### Passing message history

Reviewer 2: ```suggestion
#### Passing a Message History
```

const responseContent = response.getContent();
```

`getContent()` is a convenience method that parses the response and returns the model's output as a string.

> #### Customizing the Request Configuration

Reviewer 3: Please consider removing this section as it is the task from the other PR #188.

https://github.com/SAP/ai-sdk-js/pull/188#discussion_r1784074411

const responseContent = response.getContent();
```

`response.getContent()` is a convenience method that parses the response and returns the model's output as a string.

#### Customizing the Request Configuration

> To customize the request configuration, for example to pass additional headers to the client, use the second parameter in the `chatCompletion()` method:

Reviewer: ```suggestion
To pass custom request configuration parameters (e.g., headers) to the client, use the second parameter in the `chatCompletion()` method:
```

const responseContent = response.getContent();
```

`getContent()` is a convenience method that parses the response and returns the model's output as a string.

```ts
const finishReason = response.getFinishReason();
```

`getFinishReason()` is a convenience method that specifies the reason for stopping the chat completion request.

```ts
const tokenUsage = response.getTokenUsage();

console.log(
  `Total tokens consumed by the request: ${tokenUsage.total_tokens}\n` +
    `Input prompt tokens consumed: ${tokenUsage.prompt_tokens}\n` +
    `Output text completion tokens consumed: ${tokenUsage.completion_tokens}\n`
);
```

> `getTokenUsage()` is a convenience method that retrieves the token usage details of the chat completion request.

Reviewer: ```suggestion
```

const responseContent = response.getContent();
```

`getContent()` is a convenience method that parses the response and returns the model's output as a string.

To retrieve the `finish_reason` for stopping the chat completion request, use the convenience method `getFinishReason()`:

```ts
const finishReason = response.getFinishReason();
```

Use the `getTokenUsage()` convenience method to retrieve the token usage details of the chat completion request:

```ts
const tokenUsage = response.getTokenUsage();

console.log(
  `Total tokens consumed by the request: ${tokenUsage.total_tokens}\n` +
    `Input prompt tokens consumed: ${tokenUsage.prompt_tokens}\n` +
    `Output text completion tokens consumed: ${tokenUsage.completion_tokens}\n`
);
```

> 

Reviewer: ```suggestion
```

const responseContent = response.getContent();
```

> #### Retrieving data from the response

Reviewer 2: ```suggestion
#### Retrieving Data from the Response
```

> //Todo

Reviewer 2: [req] TODO ;)

}
```

In the above code snippet, both `orchestrationClient.chatCompletion()` and `response.getContent()` can throw errors.

- **Axios Errors**:  
  When the chat completion request fails with a `400` status code, the caught error will be an `Axios` error. The property `error.response.data.message` may provide additional details about the failure's cause.

- **Output Content Filtered**:  

>   The method `response.getContent()` can throw an error if the content filter configuration filters the output. This situation can occur even if the initial request succeeds. The `error.message` property indicates whether the output was filtered.

Reviewer: ```suggestion
  The `response.getContent()` method can throw an error if the output content filter configuration filters the output. This can occur even if the initial request succeeds. The `error.message` property indicates whether the output was filtered.
```

1. Maybe have a section for error handling? Currently this is inside content-filtering. Are axios errors limited to content filtering ?
2. What is the initial request?
3. Its not the configuration that filters the output, but the output filter.

Author: 1. As currently,  none of the clients talk about error handling, I would rather open a follow-up task that adds error handling sections to every client. I personally don't have much understanding to do proper error handling in TS, so wouldn't be able to explain it in this document now.
Axios errors are not limited to content filtering, but I hope it is clear when Axios errors occur during content filtering.
2 & 3, I have worked upon.

The Azure content filter supports four categories: `Hate`, `Violence`, `Sexual`, and `SelfHarm`.
Each category can be configured with severity levels of 0, 2, 4, or 6.

## Local Testing

> For local testing instuctions, refer [here](../../README.md#local-testing).

Reviewer 2: ```suggestion
For local testing, refer to [here](../../README.md#local-testing).
```

}
```

> In the above code snippet, both `chatCompletion()` and `getContent()` methods can throw errors.

Reviewer: ```suggestion
Both `chatCompletion()` and `getContent()` methods can throw errors.
```

- **Axios Errors**:  

>   When the chat completion request fails with a `400` status code, the caught error will be an `Axios` error. The property `error.response.data.message` may provide additional details about the failure's cause.

Reviewer: ```suggestion
  When the chat completion request fails with a `400` status code, the caught error will be an `Axios` error.
  The property `error.response.data.message` may provide additional details about the failure's cause.
```

- **Output Content Filtered**:  

>   The method `getContent()` can throw an error if the output filter filters the model output. This situation can occur even if the chat completion request succeeds with a `200` status code. The `error.message` property indicates whether the output was filtered.

Reviewer: ```suggestion
  The method `getContent()` can throw an error if the output filter filters the model output.
  This can occur even if the chat completion request responds with a `200` HTTP status code.
  The `error.message` property indicates if the output was filtered.
```

return response.getContent();
```

### Retrieving Data from the Response

> In addition to `getContent()`, other available convenience methods can retrieve the finish reason and token usage.

Author: I understand, but it is more easier to explain what `getContent()` does when you have code snippets that also support what you are trying to explain.

The error `getContent()` throws is a rather corner case and only happens during content filtering(for now).
Also, we introduce `getContent()` much before in the templating snippet, and I wanted to explain atleast a line about it there.

Would it work if I change the title to `Retrieving further information from the response` ? WDY think?

> In addition to `getContent()`, other available convenience methods can retrieve the finish reason and token usage.

Reviewer: [pp] This is just my opinion. I would prefer to show all convenience methods in one place, and right now to me it makes sense to do it when we explain `getContent`. The templating section is mandatory and the first section where basic usage is explained. I would also even just list them down, and not have a whole section for each convenience function.

> In addition to `getContent()`, other available convenience methods can retrieve the finish reason and token usage.

Author: Done now, moved the convenience methods under the templating section now rather than as a separate section.

Use `response.rawReason` to access the complete HTTP response from the orchestration service.

#### Finish Reason

> The Finish Reason indicates the reason for stopping the chat completion request.

Reviewer: ```suggestion
The finish reason indicates the reason for stopping the chat completion request.
```

> For example, when output gets filtered based on the configuration, the finish reason is `content_filter`.

Reviewer: ```suggestion
For example, when output is filtered based on the configuration, the finish reason is `content_filter`.
```

```ts
const finishReason = response.getFinishReason();
```

#### Token Usage

To retrieve the token usage details of the orchestration request, use the following snippet:

```ts
const tokenUsage = response.getTokenUsage();

logger.info(
  `Total tokens consumed by the request: ${tokenUsage.total_tokens}\n` +
    `Input prompt tokens consumed: ${tokenUsage.prompt_tokens}\n` +
    `Output text completion tokens consumed: ${tokenUsage.completion_tokens}\n`
);
```

> Remember to initialize a logger before using it.

Reviewer: Where would the user look for this information? Maybe use console.log instead?

Author: I assumed that readers of the guide would be able to look up information about logging on their own, so I didn’t feel the need to explain every detail, especially around configuring loggers.
Since console.log is often flagged by ESLint in many configurations, I thought using a logger would be a better practice.

WDY think?

Reviewer: If the idea is that we are providing code snippets that can actually be copied right away and used, then I would not introduce things for which they user has to install a library and figure out how to make it work. I would not assume they already know and have it configured in their projects. `console.log` might complain, but it works directly.
